tests/Tests/QuickCheckUtils.hs

```diff
@@ -76,54 +79,68 @@ instance Arbitrary BL.ByteString where
 -- | For tests that have O(n^2) running times or input sizes, resize
 -- their inputs to the square root of the originals.
 newtype Sqrt a = Sqrt { unSqrt :: a }
-    deriving (Eq, Show)
+ deriving (Eq, Show)
 
 instance Arbitrary a => Arbitrary (Sqrt a) where
-    arbitrary = fmap Sqrt $ sized $ \n -> resize (smallish n) arbitrary
-        where
-            smallish = round . (sqrt :: Double -> Double) . fromIntegral . abs
-    shrink = map Sqrt . shrink . unSqrt
+    arbitrary = coerce $ sized $ \n -> resize (smallish n) $ arbitrary @a
+     where
+      smallish = round . (sqrt :: Double -> Double) . fromIntegral . abs
+    shrink = map coerce . shrink @a . coerce
 
 instance Arbitrary T.Text where
-    arbitrary = (T.pack . getUnicodeString) `fmap` arbitrary
+    arbitrary = T.pack <$> listOf arbitraryUnicodeChar -- without surrogates
     shrink = map T.pack . shrink . T.unpack
 
 instance Arbitrary TL.Text where
-    arbitrary = (TL.fromChunks . map notEmpty . unSqrt) `fmap` arbitrary
+    arbitrary = TL.fromChunks <$> coerce (arbitrary @[Sqrt (NotEmpty T.Text)])
     shrink = map TL.pack . shrink . TL.unpack
 
 newtype BigInt = Big Integer
-               deriving (Eq, Show)
+ deriving (Eq, Show)
 
 instance Arbitrary BigInt where
-    arbitrary = choose (1::Int,200) >>= \e -> Big <$> choose (10^(e-1),10^e)
+    arbitrary = do
+      e <- choose (1::Int,200)
+      coerce $ choose @Integer (10^(e-1),10^e)
     shrink (Big a) = [Big (a `div` 2^(l-e)) | e <- shrink l]
-      where l = truncate (log (fromIntegral a) / log 2 :: Double) :: Integer
+     where
+      l :: Integer
+      l = truncate $ logBase @Double 2 $ fromIntegral a
 
 newtype NotEmpty a = NotEmpty { notEmpty :: a }
-    deriving (Eq, Ord, Show)
+ deriving (Eq, Ord, Show)
+
+toNotEmptyBy :: Functor m => ([Char] -> a) -> m (NonEmptyList Char) -> m (NotEmpty a)
+toNotEmptyBy f = fmap (coerce . f . coerce)
```

Comment:
E.g. here, just `fmap (coerce f)`

Reply:
Also looked at that place. The compiler does not agree to it. Conversion `[]->NonEmptyList` needs to be somewhere. `f` (`pack`) is not polymorphic to `NonEmptyList`.

There is a tradeoff. Or supply multiple the `coerce` in several places, or have it in one place there, and since `fmap (f . coerce)` is guaranteed to be free anyway.

If we were reducing `[]` from the code base - then type sig changes & `coerce` may go up the stack.

Reply:
Ahh... You are right.
Not `coerce . f`.

`coerce . f . coerce == coerce f`

Reply:
`[]->NonEmptyList` is part of `coerce`

```
coerce . f . coerce =~
coerce f . coerce =~
coerce f
```

Reply:
Yep.

I forgot that converting the data `coerce . f . coerce` is equal to converting the function/interface `coerce f`.